keps/sig-scheduling/20190317-non-preempting-priorityclass.md

```diff
@@ -42,74 +42,80 @@ Pods are be scheduled according to descending priority.
 If a pod cannot be scheduled due to insufficient resources,
 lower-priority pods will be preempted to make room.
 
-This proposal makes the preempting behavior optional for a PriorityClass,
-by adding a new field to PriorityClasses,
+This proposal makes the preemption behavior optional for a PriorityClass,
+by adding a new policy field to PriorityClasses,
 which in turn populates PodSpec.
 If a pod is waiting to be scheduled,
-and it does not have preemption enabled,
+and it does not have preempting enabled,
 it will not trigger preemption of other pods.
+While a pod is a candidate to be preempted,
+but it has the non-preemptible policy,
+it will not be taken as victim.
 
 ## Motivation
 
-Allowing PriorityClasses to be non-preempting is important for running batch workloads.
+Allowing PriorityClasses to be non-preempting/non-preemptible is important for running batch workloads.
 
 Batch workloads typically have a backlog of work,
 with unscheduled pods.
 Higher-priority workloads can be assigned a higher priority via a PriorityClass,
 but this may result in pods with partially-completed work being preempted.
-Adding the non-preempting option allows users to prioritize the scheduling queue,
+Adding the non-preempting/non-preemptible option allows users to prioritize the scheduling queue,
 without discarding incomplete work.
 
 ### Goals
 
-Add a boolean to PriorityClasses,
+Add a preemption policy filed to PriorityClasses,
 to enable or disable preemption for pods of that PriorityClass.
 
+The policy includes non-preemting and non-preemptible.
+
 ### Non-Goals
 
-* Protecting pods from preemption. PodDisruptionBudget should be used.
+* Pod high available and autoscaling. PodDisruptionBudget should be used.
 
 ## Proposal
 
-Add a Preempting field to both PodSpec and PriorityClass.
-This field will default to true,
+Add a PreemptionPolicy field to both PodSpec and PriorityClass.
+This field will default to `PreemptLowerPriority`,
 for backwards compatibility.
 
-If Preempting is true for a pod,
-the scheduler will preempt lower priority pods to schedule this pod,
-as is current behavior.
-
-If Preempting is false,
-a pod of that priority will not preempt other pods.
-
-Setting the Preempting field in PriorityClass provides a straightforward interface,
-and allows ResourceQuotas to restrict preemption.
+* PreemptLowerPriority means that pod can preempt other pods with lower priority and
+can be preempted by other pods with higher priority.
+* PreemptNever means that pod never preempts other pods with lower priority and
+can be preempted by other pods with higher priority.
+* NonPreemptible means that pod can preempt other pods with lower priority and
```

Comment:
I strongly object to a NonPreemptible policy as it clearly conflicts with high priority classes. It could cause confusion for users, is prone to mis-configuration. If cluster admins are not careful about allocating quota at low priority non-preemptible priorities, these priority classes could create abuse opportunities for users and allow them to hug cluster resources.
An example that could cause a cluster to malfunction: There are some daemon pods with critical priority in a cluster. They cannot be scheduled due to lack of resources. Today, the scheduler will remove any pod with lowest priority in the cluster to create room for such critical priority pods. Now imagine that all of those lower priority pods were `non-preemptible`. What should the scheduler do? Should it respect the critical priority of the daemons or the fact that lower priority pods are non-preemptible?

For these reasons, I believe we should only allow higher priority pods to yield to lower priority pods, but we should never let victims to decide whether they can be preempted or not.

Reply:
I understand that abusing NonPreemptible policy would cause resource problems. 

The main motivation here is to separate the preemption from priority value. For each priority class, there will be the "value" for scheduling order and "preemption policy" for preemption part.

Reply:
Thanks for your proposal, but I think this could cause issues as I explained above.